distribution/tools/server-cli/src/main/java/org/elasticsearch/server/cli/PumpThread.java

```diff
@@ -0,0 +1,129 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.server.cli;
+
+import org.elasticsearch.bootstrap.BootstrapInfo;
+import org.elasticsearch.cli.Terminal;
+
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.io.PrintWriter;
+import java.nio.charset.StandardCharsets;
+import java.util.List;
+import java.util.concurrent.CountDownLatch;
+
+import static org.elasticsearch.bootstrap.BootstrapInfo.SERVER_READY_MARKER;
+import static org.elasticsearch.server.cli.ProcessUtil.nonInterruptibleVoid;
+
+/**
+ * A thread which reads stdout or stderr of the jvm process and writes it to this process' respective stdout or stderr.
+ */
+class PumpThread extends Thread {
+    private final BufferedReader reader;
+    private final PrintWriter writer;
+
+    // an unexpected io failure that occurred while pumping stderr
+    private volatile IOException ioFailure;
+
+    private PumpThread(String name, PrintWriter output, InputStream input) {
+        super(name);
+        this.reader = new BufferedReader(new InputStreamReader(input, StandardCharsets.UTF_8));
+        this.writer = output;
+    }
+
+    void checkForIoFailure() throws IOException {
+        IOException failure = ioFailure;
+        if (failure != null) {
+            throw failure;
+        }
+    }
+
+    /**
+     * Waits for the stderr pump thread to exit.
+     */
+    void drain() throws IOException {
+        nonInterruptibleVoid(this::join);
+        checkForIoFailure();
```

Comment:
This wasn't done previously, but I think it's necessary for `ServerProcess#detach()`:
```java
/**
* Detaches the server process from the current process, enabling the current process to exit.
*
* @throws IOException If an I/O error occurred while reading stdout / stderr or closing any of the standard streams
*/
public synchronized void detach() throws IOException {
   ...
}
```